### PR TITLE
Functions for workflow update

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This Pull Request adds necessary functions to allow the update of workflows in an easy way, just by using dicts. For instance, one could add a new status and transition in the `SAMPLE_WORKFLOW` easily as follows:

```python
settings = {
    "states": {
        "published": {
            # list preserves transitions, tuple replaces
            "transitions": ["store"],
        },
        "stored": {
            "title": "Stored",
            "description": "Sample is stored",
            "transitions": ("recover", "detach", "dispatch", ),
            # Copy permissions from sample_received first
            "permissions_copy_from": "sample_received",
            # Overwrite permissions
            "permissions": {
                # Note here we are passing tuples, so these permissions are
                # set with acquire=False
                permissions.AddAnalysis: (),
                permissions.AddAttachment: (),
                permissions.TransitionCancelAnalysisRequest: (),
                permissions.TransitionReinstateAnalysisRequest: (),
                permissions.EditFieldResults: (),
                permissions.EditResults: (),
                permissions.TransitionPreserveSample: (),
                permissions.TransitionPublishResults: (),
                permissions.TransitionScheduleSampling: (),
                ModifyPortalContent: (),
            }
        },
    },
    "transitions": {
        "store": {
            "title": "Store",
            "new_state": "stored",
            "action": "Store sample",
            "guard": {
                "guard_permissions": "",
                "guard_roles": "",
                "guard_expr": "python:here.guard_handler('store')",
            }
        },
    }

}

update_workflow(SAMPLE_WORKFLOW, **settings)

```

## Current behavior before PR

Functions to update workflows easily are not present

## Desired behavior after PR is merged

Functions to update workflows easily are present

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/senaite/core/workflow/__init__.py [hunk 1]
```diff
@@ -18,10 +18,214 @@
...
+
+    # Look-up the workflow by id
+    wf_tool = api.get_tool("portal_workflow")
+    workflow = wf_tool.getWorkflowById(workflow_or_id)
+    if not workflow:
+        if default is _marker:
+            raise ValueError("Workflow not found: %s " % workflow_or_id)
+        return default
+
+    return workflow
```
Comment (ramonski) [inline src/senaite/core/workflow/__init__.py]: Would be cool to have this function accept the object as well to return the primary workflow;)

File: src/senaite/core/workflow/__init__.py [hunk 2]
```diff
@@ -18,10 +18,214 @@
...
+
+    - Transitions passed-in as a list are added to existing transitions
+    - Transitions passed-in as a tuple replace the existing transitions
+    """
+    wf = get_workflow(workflow)
+    state = get_workflow_state(wf, state_id)
+
+    # Set basic info (title, description, etc.)
+    state.title = kwargs.get("title", state.title)
+    state.description = kwargs.get("description", state.description)
```
Comment (ramonski) [inline src/senaite/core/workflow/__init__.py]: Initially I thought to pass in parameters in the following way:

`update_workflow_state(workflow, state_id, title=None, description=None, states=None, transitions=None, **kwargs)`

where the `states` and `transitions` would be the dicts with the parameters.

File: src/senaite/core/workflow/__init__.py [hunk 3]
```diff
@@ -18,10 +18,214 @@
...
+        roles = perm_info.get("roles", acquired and [] or ())
+        mapping[perm_id] = roles
+
+    # Update the permissions at destination
+    update_workflow_permissions(wf, destination_id, mapping)
+
+
+def update_workflow_transition(workflow, transition_id, **kwargs):
+    """Updates the workflow transition in accordance with settings passed-in
+    """
```
Comment (ramonski) [inline src/senaite/core/workflow/__init__.py]: same here:

`update_workflow_transition(workflow, state_id, title=None, description=None, transition=None, **kwargs)`

**Review by ramonski — CHANGES_REQUESTED** [review]
Hi Jordi,
many thanks for adding these helpers.
There are only some minor things I would like to have to make it more understandable without looking into the structure of the whole workflow dict.
I guess the changes would also allow us to add a better docstring, e.g.:

```python
def update_workflow_state(workflow, state_id, title=None, description=None, states=None, transitions=None, **kwargs):
    """Updates the state configuration of an existing workflow
       
       states = {
           "transitions": ("recover", "detach", "dispatch", ),
            # Copy permissions from sample_received first
            "permissions_copy_from": "sample_received",
            # Overwrite permissions
            "permissions": {
                # Note here we are passing tuples, so these permissions are
                # set with acquire=False
                permissions.AddAnalysis: (),
                permissions.AddAttachment: (),
                permissions.TransitionCancelAnalysisRequest: (),
                permissions.TransitionReinstateAnalysisRequest: (),
                permissions.EditFieldResults: (),
                permissions.EditResults: (),
                permissions.TransitionPreserveSample: (),
                permissions.TransitionPublishResults: (),
                permissions.TransitionScheduleSampling: (),
                ModifyPortalContent: (),
            }
       }
       
       :param state_id: Workflow ID
       :param title: Title to set (leave empty to keep the current title)
       :param description: Description to set (...)
       :param: states: Dict of new WF state configuration 

    """
```

**Comment by xispa** [comment]
> Hi Jordi, many thanks for adding these helpers. There are only some minor things I would like to have to make it more understandable without looking into the structure of the whole workflow dict. I guess the changes would also allow us to add a better docstring, e.g.:
> 

Appreciated. I agree with your suggestions, let me do these changes!

**Review by ramonski — APPROVED** [review]
Great Work!
Now we could even build a UI to change the WF permissions in the LIMS Setup:)

**Pull request merged**